### PR TITLE
linear_cross_entropy: support probability targets with reduction='none'

### PR DESCRIPTION
**Pull request by pearu**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.13.0) (oldest at bottom):
* #188433
* __->__ #188296

Extends the chunked path to probability (soft-label) targets under
reduction='none', which previously fell back to the reference. The
no_reduction op already routes through the shared accumulator (which
implements the prob math for mean/sum); this wires prob into the two
none-specific paths:

- forward (loss-only): per-sample loss s_n*log(denom_n) - <(w*t)_n,
  shifted_n>, alongside the existing index per-row loss;
- backward (recompute): the per-sample upstream grad is folded into
  prob_wt (the prob analogue of the index path's neg_weight_target), so
  the shared grad loop emits grad_output[n]*(softmax*s_n - w*t) with the
  scalar loss_scale unchanged.

chunkable_prob_target drops its reduction clause -- reduction validity is
already enforced by the surrounding conditions -- so prob targets chunk on
any reduction.

Also unifies the per-row weighting: weight_chunk becomes a property giving
the dense row mass sum_v (w*t) for prob (it was already the index per-row
weight), so the loss-only forward and the grad loop share one definition
instead of recomputing the row sum.

The none backward enters the shared grad loop only for gradients and
discards its scalar output, so it now skips the loss-only work there (the
(B, V) loss-term reduction and the two output updates) -- a Pareto win
(faster at identical memory and numerics; ~6-9% of the backward in the
small-D/large-V regime where chunking is used).

The reduction='none' ULP caps are calibrated per backend
(CPU/CUDA/MPS/ROCm) from a CI + A100 sweep; they are drift trip-wires,
with grad_error as the correctness guard. The MPS fp32 input-grad uses a
5*eps grad_error tolerance (vs the default 3*eps) -- its dense recompute
rounds to ~3.3*eps, as expected for fp32.

## Performance

reduction='none' on the chunked path is a memory feature: the materialized
reference holds the full (N, V) logits + softmax + grad-logits, while the
chunked path holds only (B, V) working buffers + the (N,) per-sample loss
(the (N, V) probability target is an input to both, a shared floor).
Measured on A100, bf16, prob target, reduction='none', swept across
num_tokens / in_features / num_classes about the anchor
(N=4096, D=2048, V=32768):

auto-dispatch prob none bf16 A100

The win is the D<<V large-vocabulary regime. At the anchor, peak memory
drops 31% -> 62% as V grows to 131072; sweeping tokens the saving grows
23% -> 69% (the (N, V) reference intermediates come to dominate the fixed
weight/target floor); sweeping in_features it shrinks from 73% at D=512 to
break-even near D=8192 and slightly negative at D=16384=V/2 (the chunk
B = N*D/V grows until chunking degenerates). The cost is throughput -- the
none backward recomputes the per-chunk forward, and the memory-optimal
chunk under-fills the GEMM -- 2.2-5.0x slower than the reference. So this
enables large-vocabulary per-sample loss under a memory budget, where the
reference would OOM; it is not a speedup. A chunking_method dial to trade
memory back for throughput is a planned follow-up.

Accuracy (bottom row of the figure) is comparable, not a clear win: fp32
logit accumulation keeps loss and input-grad (dashed) more accurate, but
the weight-grad (solid) converges toward the reference as V grows --
grad_logits = softmax - w*target is a subtraction whose per-class terms
thin to ~1/V, so its cancellation grows with V; the weight-grad sums over N
at fixed class and inherits it, while the input-grad sums over V and
averages it out. (No liger baseline -- liger's fused op is index-target
only.)

Test Plan:

    CUDA_VISIBLE_DEVICES="" python test/test_nn.py -k linear_cross_entropy -v
    python test/test_nn.py -k linear_cross_entropy -v   # CUDA (A100)

Authored with the assistance of Claude (AI coding assistant).

### CODE REVIEW

**Comment by slayton58** [comment]
@claude perform a deep review on this PR, concentrating on subtle errors that a human reviewer might miss. Only surface significant issues, ignore anything small. Reply in a collapsed comment.

**Comment by slayton58** [comment]
I think point 1) from the claude review is the most reasonable but it's not blocking.

**Review by slayton58 — APPROVED** [review]

**Comment by pearu** [comment]
Addressed issue 1; acknowledging issue 2.

**Issue 1 (missing target-grad guard, op-to-op parity):** added a guard in `_linear_cross_entropy_batch_chunked_no_reduction_backward` that raises when `needs_input_grad[2]` (the target slot) is set -- the backward never produces `result[2]`, so that is exactly where a grad-requiring probability target would otherwise silently get `None`. I used the backward guard rather than mirroring the scalar op's forward guard because inside a `@torch.library.custom_op` body `torch.is_grad_enabled()` is `False`, so the forward `requires_grad and is_grad_enabled()` shape is a no-op in eager; `needs_input_grad[2]` is reliable in both eager and traced graphs. Verified it raises on a grad-requiring prob target, does not fire on normal input/weight grads, and does not fire in inference.

**Issue 2 (none-forward einsum within-row fp16 accumulation vs the scalar loop's explicit fp32):** left as-is. The `einsum("bv,bv->b")` is per-row, so there is no batch-wide overflow; the within-row sum is `<(w*t), shifted_logits>` with `w*t >= 0` summing to ~the row mass and `shifted <= 0`, so partial sums stay in fp16 range, and the loss ULP cap guards drift. Flagging the divergence from the sibling loop as intentional.

Note: this guard was added after your approval (per the review's suggestion) -- a ~10-line defense-in-depth addition to the backward, no feature/numeric change, flagging in case you want a re-glance.

_Drafted by Claude Code (an AI agent) and reviewed & approved by pearu._
